File: day-01/main.rb
# Advent of Code 2015 - Day 1

def read_input
  File.read(File.expand_path("../input.txt", __FILE__))
end

def main
  input = read_input.tr("^()", "") # limit to valid characters

  basement = nil
  floor = 0

  input.each_char.each_with_index do |char, index|
    floor += (char == "(") ? +1 : -1
    basement = index + 1 if floor == -1 && !basement
  end

  puts "Part One: #{floor}" # final floor
  puts "Part Two: #{basement}" # index of first basement entry
end

main if "main" == to_s

File: day-01/main.py
"""Advent of Code 2015 - Day 1."""

import re
from pathlib import Path


def read_input():
    """Read input file, return string, and drop unexpected characters."""
    text = Path(__file__).with_name('input.txt').read_text()
    return re.sub(r'[^()]', '', text)


def steps_to_basement(instructions):
    """Determine number of steps needed to reach the basement."""
    floor = 0
    for step, instruction in enumerate(instructions):
        if instruction == '(':
            floor += 1
        elif instruction == ')':
            floor -= 1
        if floor == -1:
            return step + 1
    raise RuntimeError('Failed to reach basement.')


def main():
    """Main entry point of puzzle solution."""
    instructions = read_input()

    floor = instructions.count('(') - instructions.count(')')
    basement = steps_to_basement(instructions)

    print('Part One: {}'.format(floor))
    print('Part Two: {}'.format(basement))


if __name__ == '__main__':
    main()
